### ISSUE TITLE
Clarify states of worker/future and allowed operations in each state

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#139]
Currently we provide two operations on a future: wait and cancel.

We need to clarify

- what are the possible states of a worker/future
- when does it transition between those states, either as result of calling these operations or as a result of it executing

In particular, we need to be clear about what happens when you wait for a future more than one (probably a panic).

Then provide one or more functions lang.future to interrogate state.

**Comment by jclark** [comment]
We can distinguish three lifecycle stages for a worker, which always occur in this order:

1. running i.e. not terminated yet; this includes when it is waiting
2. terminated i.e. the process has terminated (either normally or abnormally) but the termination value has not been returned to another worker via wait
3. completed i.e. the process has terminated and returned its value via wait

I think waiting for a completed process should result in a panic:
- most likely a programming bug
- returning is conceptually something that is done once
- not desirable to have to keep the return value live after return
- not desirable to have the value be returned to multiple workers, which might lead to concurrent access

Apart from the 3 lifecycle stages, other things we could expose are:

- Was termination normal or abnormal? (makes sense in stage 2 or 3)
- Has a cancellation request been made? (makes sense at any stage)
- Was the termination caused by the cancellation? (makes sense at stage 2 or 3)
- Is the worker suspended? (makes sense in stage 1)

Not sure which of these we want to expose and how we want to slice them.

**Comment by jclark** [comment]
One approach is to slice the state into three parts, each returned by a separate method:

1. three values for how terminated: not yet terminated, terminated normally, terminated abnormally
2. true/false for completed (wait will panic if true)
3. true/false for cancelled (true iff the `cancel` method has been called)

**Comment by jclark** [comment]
@hasithaa Can you comment on the semantics that is currently implemented?

**Comment by jclark** [comment]
We reached two conclusions in discussion today:

- use "dead" to refer to the state where a future has been waited for 
- waiting for a dead future should result in a panic

**Comment by jclark** [comment]
@hasithaa pointed out a tricky scenario which I hadn't considered. Suppose we have a future F for a worker A that has not yet terminated.

1. Worker B waits for F
2. Worker C waits for F (F has not yet terminated so it is not dead)
3. Now worker A terminates

Current behaviour is that both waits succeed and get the termination value for F. I don't think that's the right thing to do. One of them should succeed, and one of them should fail, but how should the failing one indicate failure? A panic seems wrong, but anything else requires a different return type.

**Comment by jclark** [comment]
Apparently .NET allows you to wait for something arbitrarily many times.

This ties into static checking on workers. If we have this concept of a dead worker, then static checking on workers in a function should prevent at compile-time the possibility of multiple waits on a single worker.

**Comment by jclark** [comment]
For named workers, I think the rule should be that a named worker can be turned into a future (by accessing it as a variable) at most once.  We don't have one-to-many send-receive, so I don't think we should have one-to-many wait.

But once the named worker has been turned into a future, there is no way in general for the compiler to prevent multiple waits at compile time. One possible compromise would be to say that waiting more than once is allowed if the termination value is immutable, and otherwise results in a panic.

**Comment by jclark** [comment]
Another possibility is to do something a bit like lax typing:

* when the thing being waited for is a named worker, then checking is strict and the possibility of multiple waits is a compile-time error;
* when the thing being waited for is an expression that is not simply a named worker, then checking is lax, and multiple waits are a runtime error

**Comment by hasithaa** [comment]
I think lax like behavior is more convenient.

**Comment by jclark** [comment]
I discussed this with @sanjiva and we decided on the lax-like behavior.

**Comment by jclark** [comment]
We will deal with semantics of wait in this issue, and push functions to access state into a separate issue #316.
